diff --git i/cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_OS390.pm w/cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_OS390.pm
index 040c9a4880..74825320f1 100644
--- i/cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_OS390.pm
+++ w/cpan/ExtUtils-MakeMaker/lib/ExtUtils/MM_OS390.pm
@@ -62,7 +62,7 @@ sub xs_make_dynamic_lib {
     }
 
     push @m, sprintf <<'MAKE', $ld_run_path_shell, $self->xs_obj_opt('$@'), $dlsyms_arg, $ldfrom, $libs, $exportlist;
-	%s$(LD) %s $(LDDLFLAGS) %s $(OTHERLDFLAGS) %s $(MYEXTLIB) \
+	%s$(LD) %s $(LDDLFLAGS) %s $(OTHERLDFLAGS) $(LDFLAGS) $(LIBS) %s $(MYEXTLIB) \
 	  $(PERL_ARCHIVE) %s $(PERL_ARCHIVE_AFTER) %s \
 	  $(INST_DYNAMIC_FIX)
 	$(CHMOD) $(PERM_RWX) $@
