***************
*** 2647,2650 ****
--- 2647,2663 ----
  #endif
  
+ #ifdef __MVS__
+ static int chgfdccsid(int fd, unsigned short ccsid) {
+   attrib_t attr;
+   memset(&attr, 0, sizeof(attr));
+   attr.att_filetagchg = 1;
+   attr.att_filetag.ft_ccsid = ccsid;
+   if (ccsid != FT_BINARY) {
+     attr.att_filetag.ft_txtflag = 1;
+   }
+   return __fchattr(fd, &attr, sizeof(attr));
+ }
+ #endif
+ 
  PerlIO *
  Perl_my_popen_list(pTHX_ const char *mode, int n, SV **args)
***************
*** 2769,2773 ****
--- 2782,2794 ----
      if (did_pipes)
           PerlLIO_close(pp[0]);
+ #ifdef __MVS__
+     PerlIO* io = PerlIO_fdopen(p[This], mode);
+     if (io) {
+       chgfdccsid(p[This], 819);
+     }
+     return io;
+ #else
      return PerlIO_fdopen(p[This], mode);
+ #endif
  
  #else
***************
*** 2923,2927 ****
--- 2944,2956 ----
      if (did_pipes)
           PerlLIO_close(pp[0]);
+ #ifdef __MVS__
+     PerlIO* io = PerlIO_fdopen(p[This],	mode);
+     if (io) {
+       chgfdccsid(p[This], 819);
+     }
+     return io;
+ #else
      return PerlIO_fdopen(p[This], mode);
+ #endif
  }
  #elif defined(DJGPP)
