***************
*** 60,64 ****
  if [ "${myfirstchar}" = "23" ]; then # 23 is '#' in ASCII
    unset ebcdic
!   def_os390_cflags="$def_os390_cflags -qascii"
  else
    ebcdic=true
--- 60,66 ----
  if [ "${myfirstchar}" = "23" ]; then # 23 is '#' in ASCII
    unset ebcdic
!   def_os390_cflags="$def_os390_cflags -qascii"
!   # ensure that 'safe' putenv is used and avoid direct environ manipulation
!   def_os390_defs="$def_os390_defs -DPERL_USE_SAFE_PUTENV";
  else
    ebcdic=true
***************
*** 79,83 ****
  def_os390_cflags="$def_os390_cflags -qhaltonmsg=3296:4108 -qsuppress=CCN3159 -qfloat=ieee"
  
! def_os390_defs='-DMAXSIG=39 -DNSIG=39';     # maximum signal number; not furnished by IBM
  def_os390_defs="$def_os390_defs -DOEMVS";   # is used in place of #ifdef __MVS__
  
--- 81,85 ----
  def_os390_cflags="$def_os390_cflags -qhaltonmsg=3296:4108 -qsuppress=CCN3159 -qfloat=ieee"
  
! def_os390_defs="$def_os390_defs -DMAXSIG=39 -DNSIG=39";     # maximum signal number; not furnished by IBM
  def_os390_defs="$def_os390_defs -DOEMVS";   # is used in place of #ifdef __MVS__
  
