diff --git a/.gitignore b/.gitignore
index 309de23..1d0c3d8 100644
--- a/.gitignore
+++ b/.gitignore
@@ -56,6 +56,7 @@ perldtrace.h
 
 # general build products
 *.o
+*.d
 *.a
 *.so
 *.i
diff --git a/cpan/Encode/Makefile.PL b/cpan/Encode/Makefile.PL
index 632d736..266bb7c 100644
--- a/cpan/Encode/Makefile.PL
+++ b/cpan/Encode/Makefile.PL
@@ -133,7 +133,7 @@ WriteMakefile(
             'MORIYAMA Masayuki <msyk@mtg.biglobe.ne.jp>',
             'Nick Ing-Simmons <nick@ing-simmons.net>',
             'Nicholas Clark <nick@ccl4.org>',
-            'Olivier Mengué <dolmen@cpan.org>',
+            'Olivier Mengue <dolmen@cpan.org>',
             'otsune',
             'Pali <pali@cpan.org>',
             'Paul Marquess <paul_marquess@yahoo.co.uk>',
@@ -159,7 +159,7 @@ WriteMakefile(
             'Tony Cook <tony@develop-help.com>',
             'Vadim Konovalov <vkonovalov@peterstar.ru>',
             'Victor <victor@vsespb.ru>',
-            'Ville Skyttä <ville.skytta@iki.fi>',
+            'Ville Skytta <ville.skytta@iki.fi>',
             'Vincent van Dam <vvandam@sandvine.com>',
             'Yitzchak Scott-Thoennes <sthoenna@efn.org>',
         ],
diff --git a/doio.c b/doio.c
index 4009545..3d5dc99 100644
--- a/doio.c
+++ b/doio.c
@@ -218,57 +218,6 @@ Perl_PerlLIO_dup2_cloexec(pTHX_ int oldfd, int newfd)
 #endif
 }
 
-#if defined(OEMVS)
-  #if (__CHARSET_LIB == 1)
-#   include <stdio.h>
-#   include <stdlib.h>
-
-    static int setccsid(int fd, int ccsid) 
-    {
-      attrib_t attr;
-      int rc;
-
-      memset(&attr, 0, sizeof(attr));
-      attr.att_filetagchg = 1;
-      attr.att_filetag.ft_ccsid = ccsid;
-      attr.att_filetag.ft_txtflag = 1;
-
-      rc = __fchattr(fd, &attr, sizeof(attr));
-      return rc;
-    }
-
-    static void updateccsid(int fd, const char* path, int oflag, int perm) 
-    { 
-      int rc;
-      if (oflag & O_CREAT) {
-        rc = setccsid(fd, 819);
-      }
-    }
-
-    int asciiopen(const char* path, int oflag) 
-    {
-      int rc;
-      int fd = open(path, oflag);
-      if (fd == -1) { 
-        return fd;
-      }
-      updateccsid(fd, path, oflag, -1);
-      return fd; 
-    }
-
-    int asciiopen3(const char* path, int oflag, int perm) 
-    {
-      int rc;
-      int fd = open(path, oflag, perm);
-      if (fd == -1) { 
-        return fd;
-      }
-      updateccsid(fd, path, oflag, perm);
-      return fd;
-    } 
-  #endif
-#endif
-
 int
 Perl_PerlLIO_open_cloexec(pTHX_ const char *file, int flag)
 {
@@ -318,7 +267,7 @@ Perl_my_mkstemp_cloexec(char *templte)
 {
     int tempfd = Internal_Perl_my_mkstemp_cloexec(templte);
 #  if defined(TEMP_CCSID)
-    setccsid(tempfd, TEMP_CCSID);
+    __setccsid(tempfd, TEMP_CCSID);
 #  endif
     return tempfd;
 }
@@ -2514,6 +2463,11 @@ Perl_do_exec3(pTHX_ const char *incmd, int fd, int do_report)
             }
           doshell:
             PERL_FPU_PRE_EXEC
+#if defined(OEMVS)
+  #if (__CHARSET_LIB == 1)
+            unsetenv("_TAG_REDIR_ERR");
+  #endif
+#endif
             PerlProc_execl(PL_sh_path, "sh", "-c", cmd, (char *)NULL);
             PERL_FPU_POST_EXEC
             S_exec_failed(aTHX_ PL_sh_path, fd, do_report);
diff --git a/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm b/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
index e90b4ae..091c6e3 100644
--- a/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
+++ b/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
@@ -68,6 +68,7 @@ sub writemain{
  */
 
 #ifdef OEMVS
+#ifndef __LP64__
 #ifdef MYMALLOC
 /* sbrk is limited to first heap segment so make it big */
 #pragma runopts(HEAP(8M,500K,ANYWHERE,KEEP,8K,4K) STACK(,,ANY,) ALL31(ON))
@@ -75,6 +76,7 @@ sub writemain{
 #pragma runopts(HEAP(2M,500K,ANYWHERE,KEEP,8K,4K) STACK(,,ANY,) ALL31(ON))
 #endif
 #endif
+#endif
 
 #define PERL_IN_MINIPERLMAIN_C
 
@@ -95,6 +97,9 @@ int
 main(int argc, char **argv, char **env)
 #endif
 {
+#if defined(OEMVS) && !defined(__clang__)
+    __init();
+#endif
     int exitstatus, i;
 #ifndef NO_ENV_ARRAY_IN_MAIN
     PERL_UNUSED_ARG(env);
diff --git a/perl.h b/perl.h
index 552a106..a72a7c8 100644
--- a/perl.h
+++ b/perl.h
@@ -22,6 +22,10 @@
 #define USE_STDIO
 #endif /* PERL_FOR_X2P */
 
+#if defined(__MVS__)
+#   include "os390/os390.h"
+#endif
+
 #ifdef PERL_MICRO
 #   include "uconfig.h"
 #else
diff --git a/iperlsys.h b/iperlsys.h
index b922af0..daed107 100644
--- a/iperlsys.h
+++ b/iperlsys.h
@@ -777,21 +777,8 @@ struct IPerlLIOInfo
 #    define PerlLIO_lstat(name, buf)    PerlLIO_stat((name), (buf))
 #  endif
 #  define PerlLIO_mktemp(file)          mktemp((file))
-#  if defined(OEMVS)
-#    if (__CHARSET_LIB == 1)
-    int asciiopen(const char* path, int oflag);
-    int asciiopen3(const char* path, int oflag, int perm);
-
-#      define PerlLIO_open(file, flag)          asciiopen((file), (flag))
-#      define PerlLIO_open3(file, flag, perm)   asciiopen3((file), (flag), (perm))
-#    else
-#      define PerlLIO_open(file, flag)          open((file), (flag))
-#      define PerlLIO_open3(file, flag, perm)   open((file), (flag), (perm))
-#    endif
-#  else
-#    define PerlLIO_open(file, flag)            open((file), (flag))
-#    define PerlLIO_open3(file, flag, perm)     open((file), (flag), (perm))
-#  endif
+#  define PerlLIO_open(file, flag)            open((file), (flag))
+#  define PerlLIO_open3(file, flag, perm)     open((file), (flag), (perm))
 #  define PerlLIO_read(fd, buf, count)  read((fd), (buf), (count))
 #  define PerlLIO_rename(old, new)      rename((old), (new))
 #  define PerlLIO_setmode(fd, mode)     setmode((fd), (mode))
diff --git a/util.c b/util.c
index ca7d374..a902360 100644
--- a/util.c
+++ b/util.c
@@ -2646,22 +2646,6 @@ Perl_unlnk(pTHX_ const char *f)	/* unlink all versions of a file */
 }
 #endif
 
-#if defined(OEMVS)
-  #if (__CHARSET_LIB == 1)
-  static int chgfdccsid(int fd, unsigned short ccsid) 
-  {
-    attrib_t attr;
-    memset(&attr, 0, sizeof(attr));
-    attr.att_filetagchg = 1;
-    attr.att_filetag.ft_ccsid = ccsid;
-    if (ccsid != FT_BINARY) {
-      attr.att_filetag.ft_txtflag = 1;
-    }
-    return __fchattr(fd, &attr, sizeof(attr));
-  }
-  #endif
-#endif
-
 /*
 =for apidoc my_popen_list
 
@@ -2717,12 +2701,6 @@ Perl_my_popen_list(pTHX_ const char *mode, int n, SV **args)
         /* Close parent's end of error status pipe (if any) */
         if (did_pipes)
             PerlLIO_close(pp[0]);
-#if defined(OEMVS)
-  #if (__CHARSET_LIB == 1)
-        chgfdccsid(p[THIS], 819);
-        chgfdccsid(p[THAT], 819);
-  #endif
-#endif
         /* Now dup our end of _the_ pipe to right position */
         if (p[THIS] != (*mode == 'r')) {
             PerlLIO_dup2(p[THIS], *mode == 'r');
@@ -2798,19 +2776,7 @@ Perl_my_popen_list(pTHX_ const char *mode, int n, SV **args)
     }
     if (did_pipes)
          PerlLIO_close(pp[0]);
-#if defined(OEMVS)
-  #if (__CHARSET_LIB == 1)
-    PerlIO* io = PerlIO_fdopen(p[This], mode);
-    if (io) {
-      chgfdccsid(p[This], 819);
-    }
-    return io;
-  #else
-    return PerlIO_fdopen(p[This], mode);
-  #endif
-#else
     return PerlIO_fdopen(p[This], mode);
-#endif
 
 #else
 #  if defined(OS2)	/* Same, without fork()ing and all extra overhead... */
@@ -2889,12 +2855,6 @@ Perl_my_popen(pTHX_ const char *cmd, const char *mode)
 #define THAT This
         if (did_pipes)
             PerlLIO_close(pp[0]);
-#if defined(OEMVS)
-  #if (__CHARSET_LIB == 1)
-        chgfdccsid(p[THIS], 819);
-        chgfdccsid(p[THAT], 819);
-  #endif
-#endif
         if (p[THIS] != (*mode == 'r')) {
             PerlLIO_dup2(p[THIS], *mode == 'r');
             PerlLIO_close(p[THIS]);
@@ -2985,7 +2945,7 @@ Perl_my_popen(pTHX_ const char *cmd, const char *mode)
   #if (__CHARSET_LIB == 1)
     PerlIO* io = PerlIO_fdopen(p[This],	mode);
     if (io) {
-      chgfdccsid(p[This], 819);
+      __chgfdccsid(p[This], 819);
     }
     return io;
   #else
diff --git a/Configure b/Configure
index bd96249..53e8787 100755
--- a/Configure
+++ b/Configure
@@ -4566,7 +4566,7 @@ echo "Checking for GNU cc in disguise and/or its version number..." >&4
 $cat >try.c <<EOM
 #include <stdio.h>
 int main() {
-#if defined(__GNUC__) && !defined(__INTEL_COMPILER)
+#if defined(__GNUC__) && !defined(__INTEL_COMPILER) && !defined(__MVS__) /* xlclang sets __GNUC__ but does not set __VERSION__ */
 #ifdef __VERSION__
 	printf("%s\n", __VERSION__);
 #else
@@ -8726,10 +8726,14 @@ case "$useshrplib" in
 true)
 	case "$userelocatableinc" in
 	true|define)
-		echo "Cannot build with both -Duserelocatableinc and -Duseshrplib" >&4
-		echo "See INSTALL for an explanation why that won't work." >&4
-		exit 4
-		;;
+    case "$osname" in
+      os390) echo "Perl on z/OS uses LIBPATH to dynamically locate shared objects." ;;
+      *)
+        echo "Cannot build with both -Duserelocatableinc and -Duseshrplib" >&4
+        echo "See INSTALL for an explanation why that won't work." >&4
+        exit 4
+        ;;
+    esac
 	esac
 	case "$libperl" in
 	'')
