diff --git a/hints/os390.sh b/hints/os390.sh
index 8995d8c..97ccc57 100644
--- a/hints/os390.sh
+++ b/hints/os390.sh
@@ -18,23 +18,10 @@
 #     Mike Fulton
 #     Karl Williamson
 #
-# The z/OS 'cc' and 'ld' are insufficient for our needs, so we use c99 instead
-# c99 has compiler options specified via standard Unix-style options, but some
-# options need to be specified using -Wc,<compiler-option> or -Wl,<link-option>
 me=$0
-case "$cc" in
-'') cc='c99' ;;
-esac
-case "$ld" in
-'') ld='c99' ;;
-esac
 
 # Prepend your favorites with Configure -Dccflags=your_favorites
 
-# This overrides the name the compiler was called with.  'ext' is required for
-# "unicode literals" to be enabled
-def_os390_cflags='-qlanglvl=extc1x';
-
 # For #ifdefs in code
 def_os390_defs="-DOS390 -DZOS";
 
@@ -47,6 +34,19 @@ def_os390_defs="$def_os390_defs -D_ALL_SOURCE";
 
 case "$use64bitall" in
 '')
+  # The z/OS 'cc' and 'ld' are insufficient for our needs, so we use c99 instead
+  # c99 has compiler options specified via standard Unix-style options, but some
+  # options need to be specified using -Wc,<compiler-option> or -Wl,<link-option>
+  case "$cc" in
+  '') cc='c99' ;;
+  esac
+  case "$ld" in
+  '') ld='c99' ;;
+  esac
+
+  # This overrides the name the compiler was called with.  'ext' is required for
+  # "unicode literals" to be enabled
+  def_os390_cflags='-qlanglvl=extc1x';
   def_os390_cflags="$def_os390_cflags -qxplink"
   def_os390_cccdlflags="-qxplink"
   def_os390_ldflags="-qxplink"
@@ -54,11 +54,25 @@ case "$use64bitall" in
   def_os390_defs="$def_os390_defs -D_OE_SOCKETS";
   ;;
 *)
-  def_os390_cflags="$def_os390_cflags -Wc,lp64"
-  def_os390_cccdlflags="$def_os390_cflags -Wl,lp64"
-  def_os390_ldflags="-Wl,lp64"
+  # Use xlclang for 64-bit
+  case "$cc" in
+  '') cc='xlclang' ;;
+  esac
+  case "$ld" in
+  '') ld='xlclang' ;;
+  esac
+  def_os390_cflags="-q64"
+  def_os390_cccdlflags="$def_os390_cflags"
+  def_os390_ldflags="-q64"
 esac
 
+arch_main_objs="os390_init.o"
+archobjs="os390.o"
+
+# Help make find os390.c
+test -h os390.c || ln -s os390/os390.c os390.c
+test -h os390_init.c || ln -s os390/os390_init.c os390_init.c
+
 myfirstchar=$(od -A n -N 1 -t x $me | xargs | tr [:lower:] [:upper:] | tr -d 0)
 if [ "${myfirstchar}" = "23" ]; then # 23 is '#' in ASCII
   unset ebcdic
@@ -81,7 +95,7 @@ def_os390_cccdlflags="$def_os390_cccdlflags -qexportall"
 #          INFORMATIONAL CCN4108 ./proto.h:4534 The use of keyword '__attribute__' is non-portable.
 # 3159= Bit field type specified for &1 is not valid. Type &2 assumed.
 #       We do not care about this warning - the bit field is 1 bit and is being specified on something smaller than an int
-def_os390_cflags="$def_os390_cflags -qhaltonmsg=3296:4108 -qsuppress=CCN3159 -qfloat=ieee"
+def_os390_cflags="$def_os390_cflags -qfloat=ieee"
 
 def_os390_defs="$def_os390_defs -DMAXSIG=39 -DNSIG=39";     # maximum signal number; not furnished by IBM
 def_os390_defs="$def_os390_defs -DOEMVS";   # is used in place of #ifdef __MVS__
@@ -96,6 +110,9 @@ def_os390_defs="$def_os390_defs -DNO_LOCALE_MESSAGES"
 # Set up feature test macros required for features available on supported z/OS systems
 def_os390_defs="$def_os390_defs -D_OPEN_THREADS=3 -D_UNIX03_SOURCE=1 -D_AE_BIMODAL=1 -D_XOPEN_SOURCE_EXTENDED -D_ALL_SOURCE -D_ENHANCED_ASCII_EXT=0xFFFFFFFF -D_OPEN_SYS_FILE_EXT=1 -D_OPEN_SYS_SOCK_IPV6 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED"
 
+# Find perl base on PATH environment variable rather than hardcoding install location
+startperl='#!/bin/env perl'
+
 # Combine -D with cflags
 case "$ccflags" in
 '') ccflags="$def_os390_cflags $def_os390_defs"  ;;
@@ -234,10 +251,10 @@ esac
 # NOLOC says to use the 1047 code page, and no locale
 case "$usedl" in
 define)
-echo 'cat >.$$.c; '"$cc"' -D_OE_SOCKETS -D_ALL_SOURCE -D_SHR_ENVIRON -E -Wc,"LANGLVL(DOLLARINNAMES)",NOLOC ${1+"$@"} .$$.c | fgrep -v "??="; rm .$$.c' > cppstdin
+echo 'cat >.$$.c; '"$cc"' -D_ALL_SOURCE -D_SHR_ENVIRON -E -qdollar ${1+"$@"} .$$.c | fgrep -v "??="; rm .$$.c' > cppstdin
    ;;
 *)
-echo 'cat >.$$.c; '"$cc"' -D_OE_SOCKETS -D_ALL_SOURCE -E -Wc,"LANGLVL(DOLLARINNAMES)",NOLOC ${1+"$@"} .$$.c | fgrep -v "??="; rm .$$.c' > cppstdin
+echo 'cat >.$$.c; '"$cc"' -D_ALL_SOURCE -E -qdollar ${1+"$@"} .$$.c | fgrep -v "??="; rm .$$.c' > cppstdin
    ;;
 esac
 
