diff --git a/Makefile.SH b/Makefile.SH
index e6cb0e0..2067db7 100755
--- a/Makefile.SH
+++ b/Makefile.SH
@@ -434,6 +434,7 @@ FIRSTMAKEFILE = $firstmakefile
 
 # Any special object files needed by this architecture, e.g. os2/os2.obj
 ARCHOBJS = $archobjs
+ARCH_MAIN_OBJS = $arch_main_objs
 
 .SUFFIXES: .c \$(OBJ_EXT) .i .s .c.depends
 
@@ -586,9 +587,9 @@ miniperl_dtrace_objs = $(miniperl_objs_nodt:%=mpdtrace/%)
 perllib_dtrace_objs = $(perllib_objs_nodt:%=libpdtrace/%)
 perlmain_dtrace_objs = maindtrace/perlmain$(OBJ_EXT)
 
-miniperl_objs = $(miniperl_dtrace_objs) $(DTRACE_MINI_O)
+miniperl_objs = $(miniperl_dtrace_objs) $(DTRACE_MINI_O) $(ARCH_MAIN_OBJS)
 perllib_objs  = $(perllib_dtrace_objs) $(DTRACE_PERLLIB_O)
-perlmain_objs = $(perlmain_dtrace_objs) $(DTRACE_MAIN_O)
+perlmain_objs = $(perlmain_dtrace_objs) $(DTRACE_MAIN_O) $(ARCH_MAIN_OBJS)
 
 miniperl_dep = $(DTRACE_MINI_O)
 perllib_dep = $(DTRACE_PERLLIB_O)
@@ -599,9 +600,9 @@ perlmain_dep = $(DTRACE_MAIN_O)
 *)
     $spitshell >>$Makefile <<'!NO!SUBS!'
 
-miniperl_objs = $(miniperl_objs_nodt) $(DTRACE_MINI_O)
+miniperl_objs = $(miniperl_objs_nodt) $(DTRACE_MINI_O) $(ARCH_MAIN_OBJS)
 perllib_objs  = $(perllib_objs_nodt) $(DTRACE_PERLLIB_O)
-perlmain_objs = perlmain$(OBJ_EXT) $(DTRACE_MAIN_O)
+perlmain_objs = perlmain$(OBJ_EXT) $(DTRACE_MAIN_O) $(ARCH_MAIN_OBJS)
 
 miniperl_dep = $(miniperl_objs)
 perllib_dep = $(perllib_objs)
