***************
*** 119,125 ****
  		;;
  	os390*)
!             shrpldflags='-Wl,XPLINK,dll'
! 	    linklibperl='libperl.x'
! 	    ;;
  	esac
  	case "$ldlibpthname" in
--- 119,131 ----
  		;;
  	os390*)
!             case "$USE_64_BIT_ALL" in 
!             '')    shrpldflags='-Wl,XPLINK,dll'
! 	           linklibperl='libperl.x'
! 	           ;;
!             '')    shrpldflags='-Wl,LP64,dll'
! 	           linklibperl='libperl.x'
! 	           ;;
!             esac
!             ;;
  	esac
  	case "$ldlibpthname" in
