diff --git a/doio.c b/doio.c
index 22b0359..43b3878 100644
--- a/doio.c
+++ b/doio.c
@@ -35,7 +35,7 @@
 #endif
 #ifdef HAS_SHM
 #include <sys/shm.h>
-# ifndef HAS_SHMAT_PROTOTYPE
+#if 0 //# ifndef HAS_SHMAT_PROTOTYPE
     extern Shmat_t shmat (int, char *, int);
 # endif
 #endif
@@ -60,6 +60,10 @@
 
 #include <signal.h>
 
+#ifdef __MVS__
+#include <zos.h>
+#endif
+
 void
 Perl_setfd_cloexec(int fd)
 {
