diff --git a/cpan/IPC-SysV/SysV.xs b/cpan/IPC-SysV/SysV.xs
index 6690718aa8..b362aed17b 100644
--- a/cpan/IPC-SysV/SysV.xs
+++ b/cpan/IPC-SysV/SysV.xs
@@ -32,7 +32,7 @@
 #      include <sys/sysmacros.h>	/* SHMLBA */
 #    endif
 #    include <sys/shm.h>
-#    ifndef HAS_SHMAT_PROTOTYPE
+#    if ! defined(HAS_SHMAT_PROTOTYPE) && !defined(__MVS__)
        extern Shmat_t shmat(int, char *, int);
 #    endif
 #    if defined(HAS_SYSCONF) && defined(_SC_PAGESIZE)
